static/app/components/stream/useSelectedGroups.tsx

```diff
@@ -0,0 +1,170 @@
+import create from 'zustand';
+
+import GroupStore from 'sentry/stores/groupStore';
+
+type ComputedState = {
+  /**
+   * Are all groups currently selected?
+   */
+  allSelected: boolean;
+  /**
+   * Are any groups currently selected?
+   */
+  anySelected: boolean;
+  /**
+   * Are mulitple groups currently selected?
+   */
+  multiSelected: boolean;
+  /**
+   * The set of selected IDs
+   */
+  selectedIds: Set<string>;
+};
+
+type State = {
+  /**
+   * Add a set of group IDs. If everything is currently selecged new items will
+   * also be selected.
+   *
+   * Passing `prune` will remove existing reocrds not in the ids list.
+   */
+  add(ids: string[], prune?: boolean): void;
+  /**
+   * Computed state values
+   */
+  computed: ComputedState;
+  /**
+   * Deselect all groups
+   */
+  deselectAll(): void;
+  /**
+   * The most recently toggled item
+   */
+  lastSelected: string | null;
+  /**
+   * Mapping of selected group IDs -> boolean they are selected
+   */
+  records: Record<string, boolean>;
+  /**
+   * Trigger a selection between the last selected group and the shift selected group
+   */
+  shiftToggleItems(itemId: string): void;
+  /**
+   * Toggle selection of a group
+   */
+  toggleSelect(itemId: string): void;
+  /**
+   * Toggle selection of all groups
+   */
+  toggleSelectAll(): void;
+};
+
+const useSelectedGroups = create<State>()((set, get) => ({
+  records: {},
+  lastSelected: null,
+  computed: {
+    get selectedIds() {
+      const {records} = get();
+      return new Set(Object.keys(records).filter(id => records[id]));
+    },
+    get anySelected() {
+      return Object.values(get().records).some(selected => selected);
+    },
+    get allSelected() {
+      const selectedList = Object.values(get().records);
+      return selectedList.length > 0 && selectedList.every(selected => selected);
+    },
+    get multiSelected() {
+      return get().computed.selectedIds.size > 1;
+    },
+  },
+  add: (ids, prune) =>
+    set(({records: currentRecords, computed}) => {
+      const {allSelected} = computed;
+
+      const entries = ids
+        .filter(id => currentRecords[id] === undefined)
+        .map(id => [id, allSelected]);
+
+      // Prune existing entries if requested
+      const existingEntries = !prune
+        ? Object.entries(currentRecords)
+        : Object.entries(currentRecords).filter(([id]) => ids.includes(id));
+
+      const records = {
+        ...Object.fromEntries(existingEntries),
+        ...Object.fromEntries(entries),
+      };
+
+      return {records};
```

Comment:
I think if we're going to do `zustand` you should also trying adding the `immer` middleware which will get rid of the need for this ref manipulation to refresh state, it looks a lot closer the original store, and a more fair comparison from a code quality / simplicity standpoint. I know it's "more" since it's another concept but it helps avoid some of the problems with state mgmt in general, if we're going to switch it should realistically be to both.

eg.
```ts
  immer<State & Actions>((set) => ({
  ...
  add: (ids, prune) => 
     set(state) => {
       ids
        .filter(id => !state.records.has(id))
        .forEach(id => state.records.set(id, state.computed.allSelected));
     }
     ...
  })
  ```
  
  Alternatively, maybe give it a try as a separate PR so we can see Zustand / Zustand + Immer / vs. when discussing it in TSC.